src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,183 @@
+            catch (GracefulException)
             {
-                // Identify the intended solution folders
-                var solutionFolders = DetermineSolutionFolder(slnFile, fullProjectPath);
-
-                slnFile.AddProject(fullProjectPath, solutionFolders);
+                throw;
             }
-
-            if (slnFile.Projects.Count > preAddProjectCount)
+            catch (Exception ex) when (ex is not GracefulException)
             {
-                slnFile.Write();
+                {
+                    if (ex is SolutionException || ex.InnerException is SolutionException)
+                    {
+                        throw new GracefulException(CommonLocalizableStrings.InvalidSolutionFormatString, solutionFileFullPath, ex.Message);
+                    }
+                    throw new GracefulException(ex.Message, ex);
+                }
             }
-
-            return 0;
         }
 
-        private static IList<string> GetSolutionFoldersFromProjectPath(string projectFilePath)
+        private async Task AddProjectsToSolutionAsync(string solutionFileFullPath, IEnumerable<string> projectPaths, CancellationToken cancellationToken)
         {
-            var solutionFolders = new List<string>();
-
-            if (!IsPathInTreeRootedAtSolutionDirectory(projectFilePath))
-                return solutionFolders;
-
-            var currentDirString = $".{Path.DirectorySeparatorChar}";
-            if (projectFilePath.StartsWith(currentDirString))
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
+            SolutionModel solution = await serializer.OpenAsync(solutionFileFullPath, cancellationToken);
+            // set UTF8 BOM encoding for .sln
+            if (serializer is ISolutionSerializer<SlnV12SerializerSettings> v12Serializer)
             {
-                projectFilePath = projectFilePath.Substring(currentDirString.Length);
+                solution.SerializerExtension = v12Serializer.CreateModelExtension(new()
+                {
+                    Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)
+                });
             }
-
-            var projectDirectoryPath = TrimProject(projectFilePath);
-            if (string.IsNullOrEmpty(projectDirectoryPath))
-                return solutionFolders;
-
-            var solutionFoldersPath = TrimProjectDirectory(projectDirectoryPath);
-            if (string.IsNullOrEmpty(solutionFoldersPath))
-                return solutionFolders;
-
-            solutionFolders.AddRange(solutionFoldersPath.Split(Path.DirectorySeparatorChar));
-
-            return solutionFolders;
-        }
-
-        private IList<string> DetermineSolutionFolder(SlnFile slnFile, string fullProjectPath)
-        {
-            if (_inRoot)
+            // Set default configurations and platforms for sln file
+            foreach (var platform in new List<string>{ "Any CPU", "x64", "x86" })
```

Comment:
This was included in the previous implementation in `SlnFileExtensions` (https://github.com/dotnet/sdk/blob/e6c8e57159724c0d22640f5eadcd32954276b22f/src/Cli/dotnet/SlnFileExtensions.cs#L128C9-L151C10) but we could potentially make these a constant in case we want to modify them in the future